design_docs/rsyslog_new_input_output_roles_proposal.md

```diff
@@ -0,0 +1,255 @@
+
+For the comparison, here is the previous style, in which logs_collections (input roles) used to belong to logging_outputs (output roles) to specify the relationship between the input and the output roles.
+
+**Previous style**
+```
+logging_outputs:
+  - name: forward-output
+    type: forwards
+    rsyslog_forwards_actions:
+      - name: to-remote
+        protocol: tcp
+        target: remote_host_name.remote_domain_name
+        port: 514
+    logs_collections:
+      - name: basic-input
+        type: basics
+  - name: file-output
+    type: files
+    logs_collections:
+      - name: file-input
+        type: files
+```
+In the above example, the inputs from imjournal (basic-input) are forwarded to the remote host (forward-output); the inputs from imfile (file-input) are stored in the local file (file-output).
+
+With this style, it is not straightforward to implement one input to send multiple outputs in the logging roles.
+Moreover, to configure the input from `imjournal` to send both the remote host and the local file,
+the basics input would be placed under both forwards and files outputs as follows.
+It could give an impression basic-input0 and basic-input1 could be configured independently, but it is not.
+There is only one `imjournal` input and most likely the first one is wiped out.
+If possible, we should avoid this type of confusions in the designing phase.
+
+**Previous style example with uncertainty**
+```
+logging_outputs:
+  - name: forward-output
+    type: forwards
+    rsyslog_forwards_actions:
+      - name: to-remote
+        protocol: tcp
+        target: remote_host_name.remote_domain_name
+        port: 514
+    logs_collections:
+      - name: basic-input0
+        type: basics
+        rsyslog_imjournal_persist_state_interval: 1000
+        rsyslog_imjournal_ratelimit_interval: 10
+        rsyslog_imjournal_ratelimit_burst: 100000
+  - name: file-output
+    type: files
+    logs_collections:
+      - name: basic-input1
+        type: basics
+        rsyslog_imjournal_persist_state_interval: 3000
+        rsyslog_imjournal_ratelimit_interval: 30
+        rsyslog_imjournal_ratelimit_burst: 300000
+```
+
+### logging_outputs, logs_collections and rsyslog_use_OUTPUT_ruleset
+This is a typical example of the configuration.
+In logging_outputs, the to-be-configured output roles are listed.
+In logs_collections, the to-be-configure input roles are listed.
+Each input role has `rsyslog_use_OUTPUT_ruleset: true|false` to specify which output the log messages are sent to.
+```
+logging_outputs:
+  - name: forward-output
+    type: forwards
+    rsyslog_forwards_actions:
+      - name: to-remote
+        protocol: tcp
+        target: remote_host_name.remote_domain_name
+        port: 514
+  - name: file-output
+    type: files
```

Comment:
Can there be multiple file outputs, like this?
```
- name: file-output1
  type: files
- name: file-output2
  type: files
```
or, in general, multiple outputs with the same name? I guess at least for `forwards` it would make sense (define forwards to multiple hosts).

Reply:
> Can there be multiple file outputs, like this?
> 
> ```
> - name: file-output1
>   type: files
> - name: file-output2
>   type: files
>
> or, in general, multiple outputs with the same type? I guess at least for `forwards` it would make sense (define forwards to multiple hosts).
> ```

Actually, having multiple files declarations is not supported.  To configure multiple file outputs, you could use  `rsyslog_files_actions` like this.
https://github.com/linux-system-roles/logging/blob/master/tests/tests_files.yml#L13-L28

Forwarding also has the corresponding setting `rsyslog_forwards_actions`
https://github.com/linux-system-roles/logging/blob/master/tests/tests_forwards.yml#L14-L47

I have to confess I haven't tested this case (thanks for asking me!) to find the first file-output0 is just overwritten by file-outpu1 and no test0 and test1 actions are configured.  I will add this to the doc that output role is supposed to be one type each.
```
        logging_outputs:
          - name: file-output0
            type: files
            rsyslog_files_actions:
              - name: test0
                severity: info
                exclude: authpriv.none;auth.none;cron.none;mail.none
                path: /var/log/messages
              - name: test1
                facility: authpriv,auth
                path: /var/log/secure
          - name: file-output1
            type: files
            rsyslog_files_actions:
              - name: test2
                severity: emerg
                path: :omusrmsg:*
              - name: test3
                facility: local7
                path: /var/log/boot.log
```

Reply:
> Actually, having multiple files declarations is not supported.

What about multiple declarations for other types? Is it a general rule that every output type has only one declaration, or can there be declarations without this limitation?

Reply:
> > Actually, having multiple files declarations is not supported.
> 
> What about multiple declarations for other types?

That is supported.  This is a test case to have the files outputs and the forwards outputs.
https://github.com/nhosoi/logging/blob/combination/tests/tests_combination.yml#L19-L43
In the files outputs, 2 output files - test0 and test1 - are configured; In the forwards outputs, 2 forwards - severity_and_facility and facility_only are configured.

> Is it a general rule that every output type has only one declaration, or can there be declarations without this limitation?

The former.  It is a general rule.  To support output type specific configuration as `rsyslog_files_actions` for the files outputs and `rsyslog_forwards_action`s for the forwards outputs, we have to add the special vars.  I believe the elasticsearch outputs already have pretty good support on them.  But we need to enhance and test them.

Reply:
> > > Actually, having multiple files declarations is not supported.
> > 
> > 
> > What about multiple declarations for other types?
> 
> That is supported. This is a test case to have the files outputs and the forwards outputs.
> https://github.com/nhosoi/logging/blob/combination/tests/tests_combination.yml#L19-L43
> In the files outputs, 2 output files - test0 and test1 - are configured; In the forwards outputs, 2 forwards - severity_and_facility and facility_only are configured.

Well according to what you write below, it is a general rule that every output thype has only one declaration. Your example shows only one declaration for `files` and `forwards`.
> 
> > Is it a general rule that every output type has only one declaration, or can there be declarations without this limitation?
> 
> The former. It is a general rule. To support output type specific configuration as `rsyslog_files_actions` for the files outputs and `rsyslog_forwards_action`s for the forwards outputs, we have to add the special vars. I believe the elasticsearch outputs already have pretty good support on them. But we need to enhance and test them.

If this is the case, isn't the `name` a bit redundant? `type` seems to uniquely identify the entry. What is `name` used for?

Reply:
It could be beneficial to have `name` to make it look more like the interface of other roles, but maybe it could be allowed to omit it and the role would derive it from `type`?

Reply:
> It could be beneficial to have `name` to make it look more like the interface of other roles, but maybe it could be allowed to omit it and the role would derive it from `type`?

+1!  I will eliminate the `name` key.

Reply:
Can you please answer what is/was `name` used for before eliminating it?

Reply:
@pcahyna, I found a good use case which is a supporter of your original suggestion - using `name` for matching the input and output.  Plus we need to allow multiple instances in the output roles.

In this PR100 design, I did not allow to have multiple output instances to have separate input.  But it turned out it is essential.  This is an old configuration for elasticsearch.
```
logging_outputs:
  - name: viaq-elasticsearch
    type: elasticsearch
    logging_inputs:
      - name: viaq-input
        type: viaq
    server_host: logging-es
    <<snip>>
  - name: ovirt-elasticsearch
    type: elasticsearch
    logging_inputs:
      - name: ovirt-input
        type: ovirt
    server_host: logging-es-ovirt
    <<snip>>
```
So, what I have to do is ...
1) keeping both `name` and `type`.
2) `type` is used for the input/output type - it is the directory name in the input_roles & output_roles.
3) `name` is used for matching the input and the output.

I still have some concerns...
1) The outputs need to allow multiple instances as shown in the elasticsearch example, but the rest of the output roles do not require the ability.  (This is mainly because the files and the forwards output has rsyslog_files_action and rsyslog_forwards_action parameters.)  Is it ok to leave this asymmetry among outputs?
2) Do you think it's acceptable to enforce the `name` value has to start with the `type` value?  Like if it is the `files` type, the `name` should be `files-something`?

Then, the above example would be something like this.  (I added 'files-something' to show what we want to support in this PR.)
```
logging_outputs:
  - name: elasticsearch-viaq
    type: elasticsearch
    server_host: logging-es
    <<snip>>
  - name: elasticsearch-ovirt
    type: elasticsearch
    server_host: logging-es-ovirt
    <<snip>>
  - name: files-something
    type: files
logging_inputs:
   - name: viaq-input
     type: viaq
    rsyslog_outputs: ['elasticsearch-viaq']
  - name: ovirt-input
    type: ovirt
    rsyslog_outputs: ['files-something', 'elasticsearch-ovirt']
```

Reply:
>> Do you think it's acceptable to enforce the name value has to start with the type value? Like if it is the files type, the name should be files-something?
> I don't think this is valuable, but such value for name could be the default if name is not provided.
>
> I am still interested what is name used for currently. Does each output instance under logging_outputs correspond to a configuration file or snippet? If so, is the name attribute used to determine the name of the file? If not, what is it used for?

It is just an identifier to distinguish the action in the state file, for instance.

> Finally, it seems strange to me that there is a rsyslog_outputs attribute. All the variables are prefixed with logging_ and attributes are not namespaced. This indicates that they are in principle generic. (I know that we don't have any other provider than rsyslog currently....) IMO, the attribute should then be called just outputs because it does not seem to be more rsyslog specific than the other attributes.

+1  I'm renaming `rsyslog_outputs` to `outputs`.

I have one ansible / jinja2 question...  I'm using the `rsyslog_use_files_ruleset: true` to make `call files` available in the rsyslog config file like this.
```
          {% if element.rsyslog_use_files_ruleset | d(false) %}
          call files
          {% endif %}
```
If we switch `rsyslog_use_files_ruleset: true|false` to `outputs: ['files-A', 'elasticsearch-B', elasticsearch-C']`, we could do something like this, I guess.  But are there any nice way to avoid to have `call elasticsearch` twice in this example?  Not having `break` always makes me sad...

{% for output in element.outputs %}
  {% if output.startswith("files")
    call files
  {% elif output.startswith("elasticsearch") %}
    call elasticsearch
  {% endif %}
{% endfor %}

I have another place I need to detect the input - output match.  In the last condition in `when` in the next code, I'd like to check logging_inputs which outputs list contains the elasticsearch instance name.  When it's true, I'd like to add the logging_input name to rsyslog_elasticsearch list.  Do you think it works as expected?  (Basically, this is very close to the implementation by @sradco.)
```
- name: Set rsyslog_elasticsearch
  set_fact:
    rsyslog_elasticsearch: "{{ rsyslog_elasticsearch | d([]) }} + {{ [ item.0 | combine( {'logging_input_name': item.1.name} )] }}"
  with_subelements:
    - "{{ logging_outputs }}"
    - logging_inputs
  when:
    - item.0.type | d('') == 'elasticsearch'
    - item.1.state | d('present') == 'present'
    - item.0.name in item.1.outputs
```

Reply:
> >  I am still interested what is name used for currently. Does each output instance under logging_outputs correspond to a configuration file or snippet? If so, is the name attribute used to determine the name of the file? If not, what is it used for?

> It is just an identifier to distinguish the action in the state file, for instance.

If it is used for something, I would say it is a reason to keep it. Can you please give an example of state file with this identifier?

>  But are there any nice way to avoid to have `call elasticsearch` twice in this example? Not having `break` always makes me sad...
> 
> {% for output in element.outputs %}
> {% if output.startswith("files")
> call files
> {% elif output.startswith("elasticsearch") %}
> call elasticsearch
> {% endif %}
> {% endfor %}

I see `call elasticsearch` only once in the example.

Reply:
> I see call elasticsearch only once in the example.

Well, since element.outputs contains elasticsearch-B and elasticsearch-C, output.startwith("elasticsearch") is true twice, isn't it?  Then, `call elasticsearch` would be printed in the rsyslog config file twice...

Reply:
Ah, I see. But you should not use `output.startswith("elasticsearch")` anyway, because in `outputs`, you have names, not types, but the decision whether to include `call elasticsearch` in the file should be done according to the type, not the name. (Assuming that an output is of type `elasticsearch` if and only if it does start with "elasticsearch" does not seem like a good idea.) Therefore, I think you should convert the names to types, and remove duplicates from the list of types, and in the loop iterate over the list of types. Then in this example `call elasticsearch` will be added only once.

Reply:
If you have a list of dicts, and you need a unique list based on some key of the dict:
```yaml
---
- name: set parameters for test
  hosts: localhost
  tasks:
  - set_fact:
      logging_outputs:
      - name: name-files-foo
        type: files
      - name: name-files-bar
        type: files
      - name: name-fwd-foo
        type: fwd
      - name: name-fwd-bar
        type: fwd

  - debug:
      msg: "uniquelist = {{ logging_outputs | items2dict(key_name='type', value_name='name') }}"
```
If I run this like `ansible-playbook -vv test.yml` I get this output:
```
TASK [set_fact] ****************************************************************
task path: /home/rmeggins/linux-system-roles/kernel_settings/junk-playbook.yml:5
ok: [localhost] => {"ansible_facts": {"logging_outputs": [{"name": "name-files-foo", "type": "files"}, {"name": "name-files-bar", "type": "files"}, {"name": "name-fwd-foo", "type": "fwd"}, {"name": "name-fwd-bar", "type": "fwd"}]}, "changed": false}

TASK [debug] *******************************************************************
task path: /home/rmeggins/linux-system-roles/kernel_settings/junk-playbook.yml:16
ok: [localhost] => {
    "msg": "uniquelist = {'files': 'name-files-bar', 'fwd': 'name-fwd-bar'}"
}
```
so I'm assuming you could do something like this:
```
{% for output in logging_outputs | items2dict(key_name='type', value_name='name') %}
call {{ output }}
{% endfor %}
```
or something like that - not sure if you can use the filter directly in the `for` loop, or if you need to set the result of `items2dict` to some other intermediate variable.

Reply:
Looks like that works - I added the following task:
```yaml
  - template:
      src: junk.j2
      dest: junk.rsyslog
```
where junk.j2 is 
```
{% for output in logging_outputs | items2dict(key_name='type', value_name='name'
) %}
call {{ output }}
{% endfor %}
```
and junk.rsyslog contains
```
call files
call fwd
```